deepchem/models/tests/test_jax_model.py

```diff
@@ -87,12 +89,191 @@ def bce_loss(pred, tar, w):
 
   # JaxModel Working
   j_m = JaxModel(
-      model,
+      model.apply,
       params,
       criterion,
       output_types=['loss', 'prediction'],
       batch_size=256,
       learning_rate=0.001,
       log_frequency=2)
-  results = j_m.fit(dataset, nb_epochs=50, deterministic=True)
-  assert results < 1.0
+  _ = j_m.fit(dataset, nb_epochs=25, deterministic=True)
+  scores = j_m.evaluate(dataset, [metric])
+  assert scores[metric.name] > 0.8
+
+
+@pytest.mark.jax
+def test_overfit_subclass_model():
+  """Test fitting a JaxModel defined by subclassing Module."""
+  n_data_points = 10
+  n_features = 2
+  np.random.seed(1234)
+  X = np.random.rand(n_data_points, n_features)
+  y = (X[:, 0] > X[:, 1]).astype(np.float32)
+  dataset = dc.data.NumpyDataset(X, np.expand_dims(y, axis=1))
+
+  class Encoder(hk.Module):
+
+    def __init__(self, output_size: int = 1):
+      super().__init__()
+      self._network = hk.nets.MLP([512, 256, 128, output_size])
+
+    def __call__(self, x: jnp.ndarray):
+      x = self._network(x)
+      return x, jax.nn.sigmoid(x)
+
+  def f(x):
+    net = Encoder(1)
+    return net(x)
+
+  # Model Initilisation
+  model = hk.transform(f)
+  rng = jax.random.PRNGKey(500)
+  inputs, _, _, _ = next(iter(dataset.iterbatches(batch_size=100)))
+
+  modified_inputs = jnp.array(
+      [x.astype(np.float32) if x.dtype == np.float64 else x for x in inputs])
+  params = model.init(rng, modified_inputs)
+
+  # Loss Function
+  criterion = lambda pred, tar, w: jnp.mean(optax.sigmoid_binary_cross_entropy(pred[0], tar))
+
+  # JaxModel Working
+  j_m = JaxModel(
+      model.apply,
```

Comment:
Here it's a bit odd to pass an attribute as a parameter. Can we pass simply a model as described above and handle this internally in `JaxModel`?

Reply:
jax's core `grad` feature requires a pure function and the gradients are taken with respect to the parameters.

`model.apply` here is the pure forward function

Reply:
```
forward_fn, initilisation = hk.transform(f)
params = initilisation(rng, modified_inputs)
```
this here is the more descriptive way of writing it